### PR TITLE
fix #277565: Shortening notes too much leads to bad stem layout

### PR DESCRIPTION
**Pull request by mattmcclinch**
See https://musescore.org/en/node/277565.

Currently, the stem is only lengthened for 3, 4, or 5 hooks. This patch allows the stem to be lengthened for any number of hooks.

This patch also fixes the glyphs for 128th note flags in the mscore font so that they will align properly with the stem.

A few explanations about these changes:
- At this point in the code, the _hook property is not a reliable indication of whether or not the chord will have hooks. It is better to use hookIdx. 
- The _tab_ variable should only be non-NULL if the staff is in fact a TAB staff.
- Stems outside the staff should not be shortened so much that the hook overshoots the notehead. See _Behind Bars_, page 16.

### CODE REVIEW

**Comment by jods4** [comment]
You describe this patch as handling 3+ hooks.
I experience the same issue with only a single hook on downwards stems.
Does your patch address that?
image

**Comment by mattmcclinch** [comment]
Yes, my patch addresses that as well. It is what I meant when I was talking about stems outside the staff being shortened so much that the stem overshoots the notehead. I believe I described my patch as handling any number of hooks.

File: libmscore/chord.cpp [hunk 1]
```diff
@@ -1214,8 +1255,8 @@ qreal Chord::defaultStemLength()
       const Staff* st    = staff();
       qreal lineDistance = st ? st->lineDistance(tick()) : 1.0;
 
-      const StaffType* tab = st ? st->staffType(tick()) : 0;
-      if (tab && tab->isTabStaff()) {
+      const StaffType* tab = (st && st->isTabStaff(tick())) ? st->staffType(tick()) : 0;
```
Comment (anatoly-os) [inline libmscore/chord.cpp]: `nullptr` instead of `0`?

File: libmscore/chord.cpp [hunk 2]
```diff
@@ -1241,14 +1282,20 @@ qreal Chord::defaultStemLength()
...
+                  }
+            else {
+                  if (shortest < 3.5)
+                        shortest = 3.5;
+                  }
+            }
 
       qreal normalStemLen = small() ? 2.5 : 3.5;
-      switch (hookIdx) {
-            case 3: normalStemLen += small() ? .5  : 0.75; break; //32nd notes
```
Comment (anatoly-os) [inline libmscore/chord.cpp]: We lost the comments describing magic numbers. Could you please add more info in the comments to `hookAdjustment()` method to describe magic numbers and calculations in general?

**Comment by mattmcclinch** [comment]
I am not a fan of stem shortening to begin with, but since Elaine Gould writes about it in _Behind Bars_ (pg. 14) we have it in MuseScore. But it is a style setting that can be adjusted somewhat or even turned off completely on a per-score basis. Stems will not be shortened if the chord has a beam, which is why the beamed chords have longer stems than the non-beamed chords. Stem shortening only affects stems outside the staff, which is why the downstem chords in the picture have shorter stems than the upstem chords.

**Comment by jods4** [comment]
@mattmcclinch Thanks for the explanation! I don't like it either, so I'll make it more subte or maybe turn it off.

**Comment by lasconic** [comment]
It could be worth adding a couple of vtests for this one and make sure it works for all supported fonts.

**Comment by anatoly-os** [comment]
I will add the tests.

**Pull request merged**